vyper/semantics/validation/annotation.py

```diff
@@ -208,7 +209,10 @@ def visit_Subscript(self, node, type_):
         if isinstance(base_type, BaseTypeDefinition):
             # in the vast majority of cases `base_type` is a type definition,
             # however there are some edge cases with args to builtin functions
-            self.visit(node.slice, base_type.get_subscripted_type(node.slice.value))
+            if isinstance(base_type, MappingDefinition):
```

Comment:
this approach only works for mapping definitions. to generalize, i think it would be good to add a generic function to all sliceable types -- `get_index_type` in addition to `validate_index_type`, which returns the correct key type.